### PR TITLE
enhancement(scheduler): share waitingPods among profiles

### LINKED ISSUE #122945: Scheduler:  Share `frameworkImpl.waitingPods` among profiles
### What would you like to be added?

I hope to share the `waitingPods` among multiple profiles, instead of creating a new `waitingPods` in each profiles with the current implementation.

In other words, `waitingPods` is only instantiated once in a scheduler app.

https://github.com/kubernetes/kubernetes/blob/a1ffdedf782edf1472102b0b99c1467d4ed39753/pkg/scheduler/framework/runtime/framework.go#L253-L258

@Huang-Wei @ahg-g  PTAL, Is there any negative impact of this change that I didn't consider, thanks.

### Why is this needed?

In some scenarios, I need to traverse `all pods` waiting in the permit stage within the current scheduler app, rather than the pods that using the current profile for scheduling and waiting in the permit stage.

For example, we want use [coscheduling](https://github.com/kubernetes-sigs/scheduler-plugins/blob/master/pkg/coscheduling/README.md) to achieve `all-or-nothing`. And we have abstracted the concept of a `Cluster`, where there are two deployments under each cluster. Each deployments use different scheduler profile, and one cluster corresponds to one podGroup, as following:
image

Then, we want pods in one cluster (podA1, podA2, podB1) to be scheduled succeed/failed together. 

We assume that all pods can pass through the filter plugins, and `podA2` is the last to be scheduled. At this point, `podA1` and `podB1` are both waiting in the permit stage. Afterwards, we traverse the `waitingPods` to complete the permit waiting for `podA1` and `podB1`. However, from an implementation perspective, now we can only see `podA1` in `waitingPods`, `podB1` cannot be seen, because`podA2` & `podB1` use different scheduling profiles. Which will cause `podB1` timeout during the permit phase, even if all pods within the podGroup are successfully scheduled.

### PR DESCRIPTION
**Pull request by NoicFank**
#### What type of PR is this?
/kind feature

#### What this PR does / why we need it:
Share `waitingPods` among scheduler profiles to help us traverse all pods waiting in the permit stage within the current scheduler app, rather than the pods that using the current profile for scheduling and waiting in the permit stage.

#### Which issue(s) this PR fixes:

Fixes #122945  

#### Special notes for your reviewer:
@Huang-Wei @ahg-g PTAL, thanks.

#### Does this PR introduce a user-facing change?

```release-note
Users can traverse all the pods that are in the scheduler and waiting in the permit stage through method `IterateOverWaitingPods`. In other words,  all waitingPods in scheduler can be obtained from any profiles. Before this commit, each profile could only obtain waitingPods within that profile.
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs
NONE
```

### CODE REVIEW

**Comment by Huang-Wei** [comment]
@NoicFank could you please fix the CI failure?

**Comment by kerthcet** [comment]
The negative impact I can think of is `waitingPodsMap` is a thread-safe map, make it writeable by all profiles will aggravate the mutex-lock race.

**Comment by NoicFank** [comment]
> @NoicFank could you please fix the CI failure?

fixed. PTAL, thanks.

**Comment by NoicFank** [comment]
> The negative impact I can think of is `waitingPodsMap` is a thread-safe map, make it writeable by all profiles will aggravate the mutex-lock race.

Good thinking point, this change does have this impact, but I think the impact is limited. The more related thing is the scale of pods that are waiting and passing simultaneously during the permit stage.

**Comment by NoicFank** [comment]
UT `TestFrameworkHandler_IterateOverWaitingPods` has been added. Is there any other tests or things I can do to push this forward? @Huang-Wei @kerthcet @ahg-g

**Comment by kerthcet** [comment]
I think we need a perform test under `perf_test` to verify that there's no big performance degradation, but we lack a template yet, we should create one. Ideas @Huang-Wei ?

**Comment by kerthcet** [comment]
Ticketed https://github.com/kubernetes/kubernetes/issues/123048

**Comment by kerthcet** [comment]
But seems hard to reproduce the scenario since default scheduler doesn't have the permit plugin.

File: pkg/scheduler/scheduler_test.go [hunk 1]
```diff
@@ -19,13 +19,16 @@ package scheduler
 import (
 	"context"
 	"fmt"
+	"k8s.io/apimachinery/pkg/util/sets"
```
Comment (Huang-Wei) [inline pkg/scheduler/scheduler_test.go:22]: nit: move it down with other k8s.io imports.

**Comment by Huang-Wei** [comment]
> The negative impact I can think of is waitingPodsMap is a thread-safe map, make it writeable by all profiles will aggravate the mutex-lock race.

Practically, take coscheduling as an example, it calls `IterateOverWaitingPods()` in 3 places:

1. Permit
1. PostFilter
1. Unreserve

For Permit() and PostFilter(), they take place in a scheudling cycle, so the concurrent access can only happen when {1 & 2} competes with {3}, and {3} is pretty rare IMO. So I'm not _that_ concerned. Also, the accuracy is over perf in this case.

File: pkg/scheduler/scheduler_test.go (hunk 1 above)
Comment (NoicFank) [reply pkg/scheduler/scheduler_test.go]: fixed.

**Comment by NoicFank** [comment]
release note has been added~ Hope it's not too late.

**Review by kerthcet — COMMENTED** [review]
/lgtm
/hold cancel

Thanks!

**Comment by NoicFank** [comment]
@Huang-Wei @kerthcet  Thanks for your time and prompt reviews.

**Pull request merged**